libhb/gtk: do not modify stderr struct directly on mingw

marcosfrm opened the pull request:
Breaks mingw build when using UCRT:

```
../libhb/hb.c: In function 'redirect_thread_func': ../libhb/hb.c:2314:11: error: 'FILE' {aka 'struct _iobuf'} has no member named '_file'
 2314 |     stderr->_file = pfd[1];
      |           ^~
```

See:

https://github.com/mingw-w64/mingw-w64/commit/fe226dd9028f3f64bef4b4384edb1ef2c95c15f4

Implement the same approach used by the GTK UI code:

https://github.com/HandBrake/HandBrake/blob/3a9e37d9b211579540d63ac54c31784756986b53/gtk/src/application.c#L307-L311

**Tested on:**

- [ ] Windows 10+  (via MinGW)
- [ ] macOS 10.13+
- [ ] Ubuntu Linux
- [x] Fedora 39 mingw cross-compile (with both ucrt and msvcrt)

bradleysepos commented:
Thanks for this. Where are you getting your mingw-w64 toolchain from, Fedora or the included build script, somewhere else?

marcosfrm commented:
Official Fedora packages: `mingw64-gcc-c++` (prefix `x86_64-w64-mingw32`) and `ucrt64-gcc-c++` (prefix `x86_64-w64-mingw32ucrt`) are enough. Required dependencies are installed automatically.

https://docs.fedoraproject.org/en-US/packaging-guidelines/MinGW/

sr55 commented:
I'm afraid this doesn't work.  We get no output on this in Windows.

marcosfrm commented:
Sorry for not testing it. I was only compiling the CLI version and assumed it was working being there in the GTK code.

Found this: https://stackoverflow.com/a/25927081

This

```c
    freopen("NUL", "w", stderr);
    _dup2(pfd[1], _fileno(stderr));
```
works on both ucrt and msvcrt here. @sr55 please give it a try.

marcosfrm commented:
(I did not test the GTK part)

jstebbins commented:
The mingw port of the gtk gui rarely gets tested as far as I know. I did the port just for kicks and then pretty much set it aside. It's probably full of bit-rot.

marcosfrm commented:
With UCRT, all logging before the redirection is lost ("Compile-time hardening features are enabled" for example). It does not occur with MSVCRT somehow.

galad87 commented:
Looks good, tested on Windows and it looks fine.

jstebbins commented on libhb/hb.c:
```diff
@@ -2297,10 +2297,12 @@ static void redirect_thread_func(void * _data)
-    stderr->_file = pfd[1];
+    // Non-console windows apps do not have a stderr->_file
+    // assigned properly
+    (void) freopen("NUL", "w", stderr);
```
Question, does this leave a file named "NUL" somewhere that needs to be cleaned up?

marcosfrm replied:
It should not. `NUL` is the old DOS device which Windows will convert behind the scenes to `\Device\Null`, its `/dev/null` equivalent.

![windows-nul](https://github.com/HandBrake/HandBrake/assets/12900645/77da0ef1-0cf6-47e7-affd-ca350e7ce09c)

jstebbins reviewed (approved):
LGTM

But since this affects Windows, I'll let @sr55 make the final decision to merge

sr55 commented:
Sorry for the delayed response.

Looks like this is still causing a regression on Windows.   We are losing some (not all) log messages now, rather than them being completely missing.   As if a buffer is not being flushed correctly. 

Before then After
![image](https://github.com/HandBrake/HandBrake/assets/628593/75a2f3cc-dd35-4b8e-97ae-59019b7be150)

Note, the After on the right is missing a bunch of logging that only later appears once the encode finalises.

marcosfrm committed:
libhb/gtk: do not modify stderr struct directly on mingw

Breaks mingw build when using UCRT:

../libhb/hb.c: In function 'redirect_thread_func':
../libhb/hb.c:2314:11: error: 'FILE' {aka 'struct _iobuf'} has no member named '_file'
 2314 |     stderr->_file = pfd[1];
      |           ^~

See:

https://github.com/mingw-w64/mingw-w64/commit/fe226dd9028f3f64bef4b4384edb1ef2c95c15f4

marcosfrm commented:
Let's try `_IONBF`, set in the GTK code already.

marcosfrm commented:
Note: `setvbuf(stderr, NULL, _IONBF, 0);` called on non-Windows too.

sr55 commented:
That appears to behave better.

sr55 commented:
Merged, Thanks
